File: LevelAndroidSDK/sdk/src/main/java/com/theshopatvsp/levelandroidsdk/ble/model/ReporterConfig.java
package com.theshopatvsp.levelandroidsdk.ble.model;

import com.theshopatvsp.levelandroidsdk.ble.model.constants.ReporterType;
import com.theshopatvsp.levelandroidsdk.ble.model.constants.reporter.DataFields;
import com.theshopatvsp.levelandroidsdk.ble.model.constants.reporter.DependentDataScale;
import com.theshopatvsp.levelandroidsdk.ble.model.constants.reporter.IndependentVariableDescription;
import com.theshopatvsp.levelandroidsdk.ble.model.response.ReportAttributesData;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by andrco on 9/25/16.
 */
public class ReporterConfig implements Serializable {
    private ReporterType type;
    private int samplingHz;
    private DependentDataScale dependentDataScale;
    private IndependentVariableDescription indVarDesc;
    private int dataFields;
    private int samplesPerRecord;
    private int maxNumberOfRecords;

    public ReporterType getType() {
        return type;
    }

    public int getSamplingHz() {
        return samplingHz;
    }

    public DependentDataScale getDependentDataScale() {
        return dependentDataScale;
    }

    public IndependentVariableDescription getIndVarDesc() {
        return indVarDesc;
    }

    public int getDataFields() {
        return dataFields;
    }

    public int getSamplesPerRecord() {
        return samplesPerRecord;
    }

    public int getMaxNumberOfRecords() {
        return maxNumberOfRecords;
    }

    private ReporterConfig(ReporterType type, int samplingHz, DependentDataScale dependentDataScale,
                           IndependentVariableDescription indVarDesc, int dataFields, int samplesPerRecord, int maxNumberOfRecords) {
        this.type = type;
        this.samplingHz = samplingHz;
        this.dependentDataScale = dependentDataScale;
        this.indVarDesc = indVarDesc;
        this.dataFields = dataFields;
        this.samplesPerRecord = samplesPerRecord;
        this.maxNumberOfRecords = maxNumberOfRecords;


    }

    public static class Builder {
        private ReporterType type;
        private int samplingHz;
        private DependentDataScale dependentDataScale;
        private List<DataFields> dataFields = new ArrayList<>();
        private int samplesPerRecord;
        private int maxNumberOfRecords;
        private IndependentVariableDescription indVarDesc;

        public Builder attrs(ReportAttributesData attrs) {
            type = ReporterType.getByReporter(attrs.getReporter());
            dependentDataScale = attrs.getDepDataScale();
            samplingHz = attrs.getIndVarScale();
            samplesPerRecord = attrs.getSamplePerRecord();
            maxNumberOfRecords = attrs.getMaxRecordsPerReport();

            if (type == ReporterType.Steps) {
                this.indVarDesc = IndependentVariableDescription.SECONDS;
            } else {
                this.indVarDesc = IndependentVariableDescription.SAMPLING_HZ;
            }

            for (DataFields field : DataFields.values()) {
                if ((field.getBit() & attrs.getDataFieldsPerSample()) > 0) {
                    dataFields.add(field);
                }
            }

            return this;
        }

        public Builder step() {
            this.type = ReporterType.Steps;
            this.indVarDesc = IndependentVariableDescription.SECONDS;

            return this;
        }

        public Builder accel() {
            this.type = ReporterType.Accel;
            this.indVarDesc = IndependentVariableDescription.SAMPLING_HZ;

            return this;
        }

        public Builder gyro() {
            this.type = ReporterType.Gyro;
            this.indVarDesc = IndependentVariableDescription.SAMPLING_HZ;

            return this;
        }

        public Builder samplingFrequency(int samplingHz) {
            this.samplingHz = samplingHz;

            return this;
        }

        public Builder dependentDataScale(DependentDataScale scale) {
            this.dependentDataScale = scale;

            return this;
        }

        public Builder includeAllAxis() {
            this.dataFields.add(DataFields.INCLUDE_X_AXIS);
            this.dataFields.add(DataFields.INCLUDE_Y_AXIS);
            this.dataFields.add(DataFields.INCLUDE_Z_AXIS);

            return this;
        }

        public Builder includeMagnitude() {
            this.dataFields.add(DataFields.INCLUDE_MAGNITUDE);

            return this;
        }

        public Builder includeXAxis() {
            this.dataFields.add(DataFields.INCLUDE_X_AXIS);

            return this;
        }

        public Builder includeYAxis() {
            this.dataFields.add(DataFields.INCLUDE_Y_AXIS);

            return this;
        }

        public Builder includeZAxis() {
            this.dataFields.add(DataFields.INCLUDE_Z_AXIS);

            return this;
        }

        public Builder samplesPerRecord(int samplesPerRecord) {
            this.samplesPerRecord = samplesPerRecord;

            return this;
        }

        public Builder maxRecordsPerReport(int maxNumberOfRecords) {
            this.maxNumberOfRecords = maxNumberOfRecords;

            return this;
        }

        public ReporterConfig build() {
            int dataFieldsInt = 0;

            if (dataFields != null && !dataFields.isEmpty()) {
                for (DataFields dataField : dataFields) {
                    dataFieldsInt |= dataField.getBit();
                }
            }

            if (dependentDataScale == null) {
                dependentDataScale = DependentDataScale.ONE_TO_ONE_BIT;
            }

            return new ReporterConfig(type, samplingHz, dependentDataScale, indVarDesc, dataFieldsInt, samplesPerRecord, maxNumberOfRecords);
        }
    }
}

File: LevelIOSSdkExample/LevelIOSSdkExample/LevelIOSSdk/model/ReporterConfig.swift
//
//  ReporterConfig.swift
//  LevelIOSSdkExample
//
//  Created by Andrew Cook on 11/3/16.
//  Copyright © 2016 TheShop. All rights reserved.
//

import Foundation

class ReporterConfigBuilder {
    var reporter: ReporterType
    var samplingFrequency: Int
    var dependentDataScale: DependentDataScale
    var indVarDesc: IndependentVariableDescription
    var dataFields: [DataFields]
    var samplesPerRecord: Int
    var maxNumberOfRecords: Int
    
    init() {
        self.reporter = ReporterType.None
        self.samplingFrequency = 0
        self.dependentDataScale = DependentDataScale.OneToOneBit
        self.indVarDesc = IndependentVariableDescription.Seconds
        self.dataFields = [DataFields]()
        self.samplesPerRecord = 10
        self.maxNumberOfRecords = 10
    }
    
    func step() {
        self.reporter = ReporterType.Steps
    }
    
    func accel() {
        self.reporter = ReporterType.Accel
    }
    
    func gyro() {
        self.reporter = ReporterType.Gyro
    }
    
    func reportAttributes(attrs: ReportAttributes) {
        self.reporter = ReporterType(rawValue: attrs.reporter)!
        self.samplingFrequency = attrs.indVarScale
        self.dependentDataScale = attrs.depDataScale!
        self.indVarDesc = attrs.indVarDescription!
        self.samplesPerRecord = attrs.samplesPerRecord
        self.maxNumberOfRecords = attrs.maxRecordsPerReport
        self.dataFields = [DataFields]()
        
        for field in DataFields.cases() {
            if (attrs.dataFieldsPerSample! & field.rawValue) > 0 {
                self.dataFields.append(field)
            }
        }
    }
    
    func build() -> ReporterConfig {
        return ReporterConfig(builder: self)
    }
}

class ReporterConfig: NSObject {
    var reporter: ReporterType
    var samplingFrequency: Int
    var dependentDataScale: DependentDataScale
    var indVarDesc: IndependentVariableDescription
    var dataFields: Int
    var samplesPerRecord: Int
    var maxNumberOfRecords: Int
    
    init(builder: ReporterConfigBuilder) {
        self.reporter = builder.reporter
        self.samplingFrequency = builder.samplingFrequency
        self.dependentDataScale = builder.dependentDataScale
        self.indVarDesc = builder.indVarDesc
        
        var fields = 0
        
        for data in builder.dataFields {
            fields |= data.rawValue
        }
        
        self.dataFields = fields
        self.samplesPerRecord = builder.samplesPerRecord
        self.maxNumberOfRecords = builder.maxNumberOfRecords
    }
}
